### PR TITLE
Reduce MTD code size by removing some unused MeshCoP features

### PR DESCRIPTION
**Pull request by chshu**
* Seperate FTD-only dataset APIs;
    
* Remove some unused MeshCoP functions from MTD build;
    
* Remove some unused CLI functions from MTD build;

It reduces MTD code size by >4KB (make -f examples/Makefile-cc2650):

```bash
$ arm-none-eabi-size output/cc2650/bin/ot-cli-mtd
   text	   data	    bss	    dec	    hex	filename
  97120	    368	  12284	 109772	  1accc	output/cc2650/bin/ot-cli-mtd

$ arm-none-eabi-size output/cc2650/bin/ot-cli-mtd 
   text	   data	    bss	    dec	    hex	filename
  92592	    368	  12284	 105244	  19b1c	output/cc2650/bin/ot-cli-mtd
```

### CODE REVIEW

**Comment by chshu** [comment]
@nibanks, I add a new macro definition (OPENTHREAD_FTD) to exclude some FTD-only features from MTD build. How could I enable the macro definition for Windows APP to pass AppVeyor build? Thanks.

File: src/cli/cli.cpp [hunk 1]
```diff
@@ -133,6 +132,11 @@ const struct Command Interpreter::sCommands[] =
     { "thread", &Interpreter::ProcessThread },
     { "version", &Interpreter::ProcessVersion },
     { "whitelist", &Interpreter::ProcessWhitelist },
+
+#ifdef OPENTHREAD_FTD_CLI
+    { "delaytimermin", &Interpreter::ProcessDelayTimerMin},
+    { "joinerport", &Interpreter::ProcessJoinerPort },
+#endif
```
Comment (nibanks) [inline src/cli/cli.cpp]: I believe we want these to remain in alphabetical order, so can you just put the `#ifdef` around the original locations in the array?

File: src/cli/Makefile.am [hunk 2]
```diff
@@ -28,23 +28,43 @@
...
     cli.cpp                           \
     cli_console.cpp                   \
     cli_dataset.cpp                   \
     cli_uart.cpp                      \
     cli_udp.cpp                       \
     $(NULL)
 
+libopenthread_ftd_cli_a_CPPFLAGS    = \
+    $(CPPFLAGS_COMMON)                \
+    -DOPENTHREAD_FTD_CLI              \
```
Comment (nibanks) [inline src/cli/Makefile.am]: You can define this in [libopenthread-cli.vcxproj](https://github.com/openthread/openthread/blob/master/etc/visual-studio/libopenthread-cli.vcxproj#L42) by adding `OPENTHREAD_FTD_CLI;`.

File: src/core/meshcop/dataset_manager_ftd.cpp [hunk 3]
```diff
@@ -274,5 +274,44 @@ void PendingDataset::HandleSet(Coap::Header &aHeader, Message &aMessage, const I
...
+
+    // add pending timestamp tlv
+    mNetwork.SetTimestamp(aTimestamp);
+    HandleNetworkUpdate(flags);
+
+    // reset delay timer
+    ResetDelayTimer(kFlagNetworkUpdated);
+
+exit:
+    {}
```
Comment (aeliot) [inline src/core/meshcop/dataset_manager_ftd.cpp]: Instead of using an empty code block here, couldn't you just explicitly call `return;`?  I think that would be cleaner then `{}`.

**Comment by chshu** [comment]
The two failure unit tests (test-lowpan, test-timer) in travis check look very strange, I do not see any reason why this PR breaks the two unit tests.

They still fail even with a much simplify and no-harm commit: https://github.com/chshu/openthread/commit/79cf3c8c2743c111254613bee2d233dc488868a5
Failure log: https://travis-ci.org/chshu/openthread/jobs/204139510

I will do more investigation, and appreciate for any quick answers or suggestions, thanks.

File: src/cli/Makefile.am (hunk 2 above)
Comment (chshu) [inline src/cli/Makefile.am]: Yes, it works, thanks.

**Comment by chshu** [comment]
It seems that I need to add the new definition into unit-test Makefile.am, so then both checks could pass now. Please help to review the code, thanks.

**Comment by chshu** [comment]
I have rebased the PR to master and resolved the conflicts. Please let me know if you have any comments, thanks.

This PR removes some unused MeshCoP/Dataset related features from MTD build which helps to reduce MTD code size by more than 4KB. There are still some Router/Leader related features could be removed, I will do it later.

**Comment by chshu** [comment]
Rebased to master, no conflicts now. Please have a review, thanks.

**Comment by jwhui** [comment]
@chshu, now that we've merged #1620, would you like to start moving this PR forward again?

**Comment by chshu** [comment]
@jwhui, Ok, sure, I will rebase it to latest master, thanks.

**Comment by chshu** [comment]
It is rebased to latest master, please help to review, thanks.

File: include/openthread/thread_ftd.h [hunk 4]
```diff
@@ -28,27 +28,61 @@
...
  * @file
- *   This file includes definitions for a MeshCoP Leader.
+ * @brief
+ *  This file defines the OpenThread Thread API (FTD only).
  */
 
-#ifndef MESHCOP_LEADER_HPP_
-#define MESHCOP_LEADER_HPP_
+#ifndef OPENTHREAD_THREAD_FTD_H_
+#define OPENTHREAD_THREAD_FTD_H_
```
Comment (DuaneEllis-TI) [inline include/openthread/thread_ftd.h]: The include guard here does not seem to match the filename.
Any particular reason?

File: src/core/meshcop/leader.hpp [hunk 5]
```diff
@@ -26,11 +26,131 @@
...
+#include <thread/mle.hpp>
+
+namespace Thread {
+namespace MeshCoP {
+
+OT_TOOL_PACKED_BEGIN
+class CommissioningData
+{
+public:
+    uint8_t GetLength(void) const {
```
Comment (DuaneEllis-TI) [inline src/core/meshcop/leader.hpp]: I mention this because an earlier theme I saw was moving getter/setter operations into the C++ headers.

Thus if this member function has the potential to be used more then once, then this should be a a CPP file not the header file. My point is this, using a get function as an example, is about 2 opcodes. Often it is a single opcode because a register has been previously allocated to the "this" pointer

```
    ldr  r0,_this_pointer
    str  r1,[r0,#member_offset]
```
That is almost the same cost as a subroutine call. The advantage is the compiler can optimize register allocation across that inlined getter/setter function.

This member function has potentially 3 subroutine calls.

These little things add up...  Death by many paper cuts.

Comment (jwhui) [inline src/core/meshcop/leader.hpp]: @abtink, I believe you made a move towards inlining get/set methods in #1550.  Is @DuaneEllis-TI's comment above consistent with your findings?

File: include/openthread/thread_ftd.h [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- *  Copyright (c) 2016, The OpenThread Authors.
+ *  Copyright (c) 2017, The OpenThread Authors.
```
Comment (DuaneEllis-TI) [inline include/openthread/thread_ftd.h]: nit: Should this be: 2016 - 2017 reflecting that there is/was a copyright in 2016?

File: include/openthread/thread_ftd.h (hunk 4 above)
Comment (DuaneEllis-TI) [inline include/openthread/thread_ftd.h]: Oops my bad!

File: src/cli/cli.cpp [hunk 7]
```diff
@@ -46,6 +46,8 @@
 #endif
 
 #include "openthread/openthread.h"
+#include "openthread/dataset_ftd.h"
```
Comment (DuaneEllis-TI) [inline src/cli/cli.cpp]: Does this require a #if OPENTHREAD_FTD guard?

File: src/core/thread/thread_netif.cpp [hunk 8]
```diff
@@ -92,8 +92,12 @@ ThreadNetif::ThreadNetif(Ip6::Ip6 &aIp6):
 #if OPENTHREAD_ENABLE_JAM_DETECTION
     mJamDetector(*this),
 #endif // OPENTHREAD_ENABLE_JAM_DETECTTION
+
```
Comment (DuaneEllis-TI) [inline src/core/thread/thread_netif.cpp]: nit: blank lines before/after are not like others here.

File: src/core/meshcop/joiner_router.hpp [hunk 9]
```diff
@@ -26,10 +26,230 @@
...
+ *
+ */
+class DelayedJoinEntHeader
+{
+public:
+    /**
+     * Default constructor for the object.
+     *
+     */
+    DelayedJoinEntHeader(void) { memset(this, 0, sizeof(*this)); }
```
Comment (aeliot) [inline src/core/meshcop/joiner_router.hpp]: Could this be split up so that the braces are on their own lines?

File: src/core/meshcop/joiner_router.hpp [hunk 10]
```diff
@@ -26,10 +26,230 @@
...
+
+    /**
+     * This constructor initializes the object with specific values.
+     *
+     * @param[in]  aSendTime     Time when the message shall be sent.
+     * @param[in]  aDestination  IPv6 address of the message destination.
+     *
+     */
+    DelayedJoinEntHeader(uint32_t aSendTime, Ip6::MessageInfo &aMessageInfo, const uint8_t *aKek) {
```
Comment (aeliot) [inline src/core/meshcop/joiner_router.hpp]: Brace should be on its own line.

File: include/openthread/thread_ftd.h (hunk 6 above)
Comment (jwhui) [inline include/openthread/thread_ftd.h]: Agree.  Should be "2016-2017" (no spaces).

Comment (chshu) [inline include/openthread/thread_ftd.h]: Actually this is a new file, although github show it as modification. Could I confirm that should we use "2017" or "2016-2017" here?

File: src/cli/cli.cpp (hunk 7 above)
Comment (chshu) [inline src/cli/cli.cpp]: Yes, updated, thanks.

File: src/core/meshcop/joiner_router.hpp (hunk 9 above)
Comment (chshu) [inline src/core/meshcop/joiner_router.hpp]: In current OpenThread header files, it seems that it is not a rule to put braces on their own lines. I think we should address this kind of format updates together in a separate PR (if we decide to follow that rule), do you agree?

File: src/core/meshcop/joiner_router.hpp (hunk 10 above)
Comment (chshu) [inline src/core/meshcop/joiner_router.hpp]: we should address this kind of format updates together in a separate PR (if we decide to follow that rule), do you agree?

File: src/core/thread/thread_netif.cpp (hunk 8 above)
Comment (chshu) [inline src/core/thread/thread_netif.cpp]: updated, thanks.

File: include/openthread/thread_ftd.h (hunk 6 above)
Comment (jwhui) [inline include/openthread/thread_ftd.h]: Given that code in this file was under the 2016 copyright, I think we should make it "2016-2017".

File: src/core/meshcop/joiner_router.hpp (hunk 9 above)
Comment (jwhui) [inline src/core/meshcop/joiner_router.hpp]: No need to worry about this here.  This will be addressed as part of #1560.

File: src/core/meshcop/joiner_router.hpp (hunk 10 above)
Comment (jwhui) [inline src/core/meshcop/joiner_router.hpp]: No need to worry about this here.  This will be addressed as part of #1560.

File: include/openthread/thread_ftd.h (hunk 6 above)
Comment (chshu) [inline include/openthread/thread_ftd.h]: Ok, got it. I have updated them, thanks.

File: src/core/meshcop/leader.hpp (hunk 5 above)
Comment (abtink) [inline src/core/meshcop/leader.hpp]: The change in #1550 reduced the MTD build size by about 2.2KB (on an internal hw platform). 

That said, I fully agree with @DuaneEllis-TI. We should consider inlining only simple methods. #1550 inlined the simple getters/setters which were basically doing member variable access.

Comment (chshu) [inline src/core/meshcop/leader.hpp]: This method is used only once: [here](https://github.com/openthread/openthread/blob/master/src/core/meshcop/leader.cpp#L109), FTD build size will **increase** 32 bytes after moving it from head file to CPP file:

- keep it in head file as is:
```
$ arm-none-eabi-size output/cc2538/bin/ot-cli-ftd 
   text	   data	    bss	    dec	    hex	filename
 143668	   4208	  17404	 165280	  285a0	output/cc2538/bin/ot-cli-ftd
```
- move it to CPP file: 
```
$ arm-none-eabi-size output/cc2538/bin/ot-cli-ftd 
   text	   data	    bss	    dec	    hex	filename
 143700	   4208	  17404	 165312	  285c0	output/cc2538/bin/ot-cli-ftd
```
So I think we should keep it as is, please let me know if you have any other concerns.

Comment (DuaneEllis-TI) [inline src/core/meshcop/leader.hpp]: Understood - it technically better this way

If the policy is generally: Simple Getter/Setter is always inlined == And all other implementations should be done in an implementation file, then this does not follow that policy.

Developers are mostly"lemmings" and write via cut/paste techniques, they may/will question - why do we sometimes do XXX, and other times not do XXX - 

You know this, you determined this because you have deep knowledge about this and experimented - others do not have your depth here (ie: none of us experimented)

Going forward, I don't think we want to have this level of back and forth for every inlined C++ function.

Bottom line - I defer to others about the final call here.

Comment (jwhui) [inline src/core/meshcop/leader.hpp]: FWIW, the current [STYLE_GUIDE.md](https://github.com/openthread/openthread/blob/master/STYLE_GUIDE.md) contains the following statement:

> - Inline functions should be used judiciously.
>   - The use of code in headers and, more specifically, the use of the non-local scope inline functions should be avoided. Exception: Simple setters and getters are fine since the compiler can efficiently optimize these and make their overhead as low as a direct data member access.

I agree this probably assumes that developers are always striving to do the right thing and performing the kind of evaluation @chshu has done above.
